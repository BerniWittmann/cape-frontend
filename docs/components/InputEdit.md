# InputEdit

A component to edit a text but hidden behind a text with an edit button

## Props

<!-- @vuese:InputEdit:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|value|The Value of the input field|`String`|`true`|-|
|rules|The Rules array that the input field should be validated against|`Array`|`false`||
|size|Size of the controls|`'medium'` / `'small'` / `'mini'`|`false`|'normal'|
|placeholder|The Placeholder for the input field|`String`|`false`|undefined|
|propName|The name the field should be referenced with within the form|`String`|`false`|'text'|

<!-- @vuese:InputEdit:props:end -->


## Events

<!-- @vuese:InputEdit:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|change|Fired when the text changes|The new text|

<!-- @vuese:InputEdit:events:end -->


## Methods

<!-- @vuese:InputEdit:methods:start -->
|Method|Description|Parameters|
|---|---|---|
|showInput|shows the Input field|-|

<!-- @vuese:InputEdit:methods:end -->


