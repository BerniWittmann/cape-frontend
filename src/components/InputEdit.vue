<template>
  <div class="input-edit">
    <span v-if="!isEditing">{{ data.text }}</span>
    <span v-else>
      <el-form ref="textForm" :model="data" label-position="top" :rules="validationRules" status-icon inline :size="size"
               @submit.native.prevent="hideInput" class="form">
        <el-form-item :prop="propName">
          <el-input
                  v-model="data.text"
                  ref="inputField"
                  :size="size"
                  :placeholder="placeholder"
                  @blur="hideInput"
                  @keyup.enter.native="hideInput"
                  @submit.native.prevent="hideInput"
                  class="input"
          ></el-input>
        </el-form-item>
      </el-form>
    </span>
    <el-button
            type="text"
            icon="el-icon-edit"
            @click="showInput"
            v-if="!isEditing"
            class="black-color tag-space"
    ></el-button>
  </div>
</template>

<script>
/* ============
 * Input Edit Component
 * ============
 *
 * A component to edit aa text but hidden behind a text with an edit button
 */

export default {
  model: {
    prop: 'value',
    event: 'change'
  },

  props: {
    value: {
      type: String,
      required: true
    },
    rules: {
      type: Array,
      default: () => []
    },
    size: {
      type: String,
      default: 'normal'
    },
    placeholder: {
      type: String,
      default: undefined
    },
    propName: {
      type: String,
      default: 'text'
    }
  },

  data() {
    return {
      data: {
        text: undefined
      },
      isEditing: false
    }
  },

  computed: {
    validationRules() {
      return {
        [this.propName]: this.rules
      }
    }
  },

  watch: {
    value: {
      deep: true,
      immediate: true,
      handler: function (newValue) {
        this.data.text = newValue
      }
    }
  },

  methods: {
    showInput() {
      this.isEditing = true
      this.$nextTick(() => {
        try {
          this.$refs.inputField.$refs.input.focus()
        } catch (e) {}
      })
    },

    hideInput() {
      this.$refs.textForm.validate((valid) => {
        if (valid) {
          this.isEditing = false
        }
      })
    }
  }
}

</script>

<style scoped lang="scss">
.input-edit {
  display: inline;
}
.form {
  margin-top: 10px;
  display: inline;
}
.black-color {
  color: #000000;
}

.input {
  width: 300px;
}

.tag-space {
  margin-left: 5px;
}
</style>
