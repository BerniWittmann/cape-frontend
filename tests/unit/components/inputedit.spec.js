import { mount } from '@vue/test-utils'
import { i18n } from '../setupPlugins'

import InputEdit from '@/components/InputEdit.vue'

describe('Components', () => {
  describe('InputEdit', () => {
    let propsData
    let cmp

    beforeEach(() => {
      propsData = {
        value: 'Test Text',
        rules: [
          { required: true, message: 'Text is required!', trigger: 'blur' }
        ]
      }
      console.warn = jest.fn()
      render()
    })

    function render() {
      cmp = mount(InputEdit, {
        i18n,
        propsData
      })
    }

    it('renders', () => {
      expect(cmp.html()).toMatchSnapshot()
    })

    it('shows the text', () => {
      expect(cmp.text()).toContain('Test Text')
    })

    it('has an edit button', () => {
      const btn = cmp.find('button')
      expect(btn.exists()).toBeTruthy()
    })

    describe('it can edit the text', () => {
      beforeEach(() => {
        const btn = cmp.find('button')
        expect(btn.exists()).toBeTruthy()

        btn.trigger('click')
      })

      it('hides the edit button', () => {
        const btn = cmp.find('button')
        expect(btn.exists()).toBeFalsy()
      })

      it('shows an input field ', () => {
        expect(cmp.html()).toMatchSnapshot()
        expect(cmp.contains('form')).toBeTruthy()
        expect(cmp.contains('input')).toBeTruthy()
      })

      it('can change the text', () => {
        const input = cmp.find('input')
        input.setValue('My New Text')
        input.trigger('blur')

        expect(cmp.html()).toMatchSnapshot()
        expect(cmp.text()).toContain('My New Text')
      })

      it('hides the input on blur', () => {
        const input = cmp.find('input')
        input.setValue('My New Text')
        input.trigger('blur')

        expect(cmp.html()).toMatchSnapshot()
        expect(cmp.contains('form')).toBeFalsy()
        expect(cmp.contains('input')).toBeFalsy()
        const btn = cmp.find('button')
        expect(btn.exists()).toBeTruthy()
      })

      it('validates the input', (done) => {
        const input = cmp.find('input')
        input.setValue('')
        input.trigger('blur')

        expect(cmp.contains('form')).toBeTruthy()
        expect(cmp.contains('input')).toBeTruthy()
        const btn = cmp.find('button')
        expect(btn.exists()).toBeFalsy()

        cmp.vm.$nextTick(() => {
          expect(cmp.html()).toMatchSnapshot()
          const message = cmp.find('.el-form-item__error')
          expect(message.exists()).toBeTruthy()
          expect(message.text()).toEqual('Text is required!')
          done()
        })
      })

      it('does not validate the input if no rule given', (done) => {
        propsData.rules = undefined
        render()
        const button = cmp.find('button')
        expect(button.exists()).toBeTruthy()

        button.trigger('click')

        const input = cmp.find('input')
        input.setValue('')
        input.trigger('blur')

        expect(cmp.contains('form')).toBeFalsy()
        expect(cmp.contains('input')).toBeFalsy()
        const btn = cmp.find('button')
        expect(btn.exists()).toBeTruthy()

        cmp.vm.$nextTick(() => {
          expect(cmp.html()).toMatchSnapshot()
          const message = cmp.find('.el-form-item__error')
          expect(message.exists()).toBeFalsy()
          expect(cmp.text()).not.toContain('Test Text')
          done()
        })
      })
    })
  })
})
