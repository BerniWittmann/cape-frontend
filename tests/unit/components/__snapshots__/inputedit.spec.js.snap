// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components InputEdit it can edit the text can change the text 1`] = `
<div class="input-edit"><span>My New Text</span> <button type="button" class="el-button black-color tag-space el-button--text">
    <!----><i class="el-icon-edit"></i>
    <!----></button></div>
`;

exports[`Components InputEdit it can edit the text does not validate the input if no rule given 1`] = `
<div class="input-edit"><span></span> <button type="button" class="el-button black-color tag-space el-button--text">
    <!----><i class="el-icon-edit"></i>
    <!----></button></div>
`;

exports[`Components InputEdit it can edit the text hides the input on blur 1`] = `
<div class="input-edit"><span>My New Text</span> <button type="button" class="el-button black-color tag-space el-button--text">
    <!----><i class="el-icon-edit"></i>
    <!----></button></div>
`;

exports[`Components InputEdit it can edit the text shows an input field  1`] = `
<div class="input-edit"><span><form class="el-form form el-form--label-top el-form--inline"><div class="el-form-item el-form-item--feedback is-required el-form-item--normal"><!----><div class="el-form-item__content"><div class="input el-input el-input--normal"><!----><input type="text" autocomplete="off" class="el-input__inner"><!----><!----><!----></div><!----></div></div></form></span>
  <!---->
</div>
`;

exports[`Components InputEdit it can edit the text validates the input 1`] = `
<div class="input-edit"><span><form class="el-form form el-form--label-top el-form--inline"><div class="el-form-item el-form-item--feedback is-error is-required el-form-item--normal"><!----><div class="el-form-item__content"><div class="input el-input el-input--normal"><!----><input type="text" autocomplete="off" class="el-input__inner"><!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><!----></span><i class="el-input__icon el-input__validateIcon el-icon-circle-close"></i></span>
  <!---->
</div>
<div class="el-form-item__error" name="el-zoom-in-top">
  Text is required!
</div>
</div>
</div>
</form></span>
<!---->
</div>
`;

exports[`Components InputEdit renders 1`] = `
<div class="input-edit"><span>Test Text</span> <button type="button" class="el-button black-color tag-space el-button--text">
    <!----><i class="el-icon-edit"></i>
    <!----></button></div>
`;
